### ISSUE TITLE
Log PROD/CSD tags from Page Curation

### ISSUE THREAD

**Issue by MusikAnimal** [azatoth/twinkle#709]
See https://phabricator.wikimedia.org/T230455 for context.

[Page Curation](https://en.wikipedia.org/wiki/Wikipedia:Page_Curation) is a MediaWiki tool to help with new page patrolling, and offers some functionality similar to Twinkle. Users of the extension have requested that speedy/PROD nominations from Page Curation be logged to Special:MyPage/CSD_log, etc., just as Twinkle does. This was investigated and decided that this is solely a Twinkle invention, and especially given the changes to the logic over the years, it is better handled by Twinkle itself rather than MediaWiki.

https://gerrit.wikimedia.org/r/530977 introduces an "action queue" (similar to mw.hook) that allows integration with Page Curation. The difference from mw.hook being that the action queue is promised-based, which is necessary since Page Curation wants to reload the page after tagging. The proposal is for Twinkle to listen to the `delete` event and log the nominations just as Twinkle's speedy module does.

I don't think this will be a lot of work, and it'd really help out the NPP community. I'm happy to take it on, if no one is up for it, but I'd like to get the go ahead from you all.

The action queue functionality goes live next Thursday, September 26. Documentation for action queue is at https://www.mediawiki.org/wiki/Extension:PageTriage#Client-side_hooks

**Comment by Amorymeltzer** [comment]
>I'm happy to take it on

Go for it!  I'll gladly review.

Are you thinking to place this in the csd module, or something separate?  It's a bit of a departure from most of Twinkle, but it seems like something that'd help (reviewing) NPP work, and anything we can do to help them is probably worth it.

**Comment by MusikAnimal** [comment]
> Go for it! I'll gladly review.

Thanks :)

> Are you imagining this living in the csd module, prod, or something new/separate? It's a bit of a departure from most of Twinkle, but it seems like something that'd help (reviewing) NPP work, and anything we can do to help them is probably worth it.

Yes I was thinking the speedy module, where it currently handles the logging. We'd register the hook (action callback, technically is I guess is what it's called) somewhere towards the bottom, with clear comments as to why it's there. The NPP community wanted Page Curation to go off of twinkleoption.js, too, so it was obvious that Twinkle should be doing the work. I'm glad to hear approval :) `Twinkle.speedy.callbacks.user.addToLog` should work nicely. We _might_ need to adjust some code to have it return a promise, but that shouldn't be too hard either.

**Comment by siddharthvp** [comment]
For whatever my opinion is worth, I don't think this should be a part of Twinkle. I'd say create a separate gadget (available only to new page reviewers) that logs the nomination details,  after reading the userspace log page names from twinkleoptions.js. 

Twinkle too can be used for speedy/PROD nominations, so I don't see why Twinkle should help out its competitor to gain ground. TW is better in most aspects. PC's PROD module doesn't do checks for previous PROD/XFD nominations, nor does it add {{old prod}} to the talk page to allow for TW to perform this check later. See the [latest NPR newsletter](https://en.wikipedia.org/wiki/User_talk:SD0001#New_Page_Review_newsletter_September-October_2019) where Mr. Self-Appointed-Coordinator-Emeritus "strongly encourages" the ~680 subscribers to _not_ use Twinkle, by citing non-reasons.  And also their comments at [WP:PCSI](https://en.wikipedia.org/wiki/WP:PCSI) where they express their desire to "wean" users away from Twinkle and other scripts and make them use PageCuration exclusively. This IMHO is not a good thing since no one will probably be improving PageCuration once this wishlist-based work period ends. Thankfully, looking at [Special:RecentChangesLinked/User:DumbBOT/ProdSummary](https://en.wikipedia.org/wiki/Special:RecentChangesLinked?hidebots=1&hidecategorization=1&hideWikibase=1&target=&limit=500&days=30&urlversion=2) (showing a 69/12 ratio of TW/PC usage for PRODing), [Special:RecentChangesLinked/CAT:CSD](https://en.wikipedia.org/wiki/Special:RecentChangesLinked?hidebots=1&hidecategorization=1&hideWikibase=1&target=Category%3ACandidates_for_speedy_deletion&limit=500&days=30&urlversion=2) (10/3 TW/PC ratio) and [history of yesterday's AFD log](https://en.wikipedia.org/w/index.php?title=Wikipedia:Articles_for_deletion/Log/2019_September_20&offset=&limit=500&action=history) (35/4 ratio), it seems that not much damage has been done.

**Comment by Amorymeltzer** [comment]
Let's see what it looks like.  While I certainly prefer Twinkle usage, it's not a competition.  In general, I think it's better not to insert Twinkle to places outside of Twinkle, but the alternative here is having all NPPers load in a new, separate gadget or userscript, whereas they already have Twinkle installed.  In this regard, the existence of CSD/PROD logs is a Twinkle invention, so it does make some sense to ensure we're the "owners" of those pages/preferences.  At any rate, let's see what it looks like.

**Comment by MusikAnimal** [comment]
Yes, I totally understand where you're coming from @siddharthvp, but I also don't think we need to worry too much about Twinkle losing market share :) It's well-understood that Page Curation is geared specifically toward new pages, and offers some features that Twinkle is unable to (Special:NewPagesFeed and associated metadata in the curation toolbar), and Twinkle meanwhile is a general purpose Swiss army knife that no one is going to uninstall. They wanted to go off of twinkleoptions.js, after all, so that secures Twinkle as a prerequisite.

The bottom line is relying on a gadget from within MediaWiki isn't really an option. The community really wanted the CSD/PROD logging. Allowing Twinkle to integrate with Page Curation seemed like the best compromise, rather than trying to reinvent the wheel and create a maintenance burden, which as you correctly say won't scale once development stalls.

The action queue functionality has been deployed so I will get a start on this soon. Thanks!

**Comment by siddharthvp** [comment]
@MusikAnimal It's just that I have never understood why PC's tagging and deletion modules exist. Those who request features for it, as well those who work on it, both seem to be trying to make it "on par with Twinkle" (that being a commonly used phrase, re WP:PCSI). Heck, why not just ask people to use Twinkle instead, and restrict PC to its newpagesfeed/metadata/wikilove etc modules? IMHO, lots of paid developer time that could have been better utilized in building software that doesn't duplicate existing ones.

**Comment by Samwalton9** [comment]
@MusikAnimal Do you have any interest in taking this on? I know this is still one of the top requests from NPP and it seems like this is the path that gets us to a solution quickest (versus doing this in PageTriage).

**Comment by MusikAnimal** [comment]
> @MusikAnimal Do you have any interest in taking this on? I know this is still one of the top requests from NPP and it seems like this is the path that gets us to a solution quickest (versus doing this in PageTriage).

I don't think I have time right now, unfortunately. I haven't touched Twinkle in ages.

For anyone interested, docs on integrating with PageTriage are at https://www.mediawiki.org/wiki/Extension:PageTriage#Client-side_hooks
